packages/gatsby-source-filesystem/src/create-file-node.js

```diff
@@ -1,82 +1,75 @@
 const prettyBytes = require(`pretty-bytes`)
+const slash = require(`slash`)
 
-const md5File = require(`bluebird`).promisify(require(`md5-file`))
-const crypto = require(`crypto`)
+const createId = src => `${slash(src)} absPath of file`
 
-const createId = path => {
-  const slashed = slash(path)
-  return `${slashed} absPath of file`
-}
+const createFileNode = async (src, opts = {}) => {
+  let node = {
+    absolutePath: slash(src),
+    children: [],
+    cwd: slash(opts.path || process.cwd()),
+    id: createId(src),
+    parent: `___SOURCE___`,
+    sourceInstanceName: opts.name || `__PROGRAMATTIC__`,
+  }
 
-exports.createId = createId
+  const stats = await fs.stat(node.absolutePath)
+
+  node = {
+    ...node,
+    ...path.parse(node.absolutePath),
+    ...stats,
+    absPath: slash(path.resolve(node.absolutePath)),
+    relativePath: slash(path.relative(node.cwd, node.absolutePath)),
+  }
 
-exports.createFileNode = async (pathToFile, pluginOptions = {}) => {
-  const slashed = slash(pathToFile)
-  const parsedSlashed = path.parse(slashed)
-  const slashedFile = {
-    ...parsedSlashed,
-    absolutePath: slashed,
-    // Useful for limiting graphql query with certain parent directory
-    relativeDirectory: path.relative(
-      pluginOptions.path || process.cwd(),
-      parsedSlashed.dir
-    ),
+  node = {
+    ...node,
+    accessTime: node.atime,
+    birthTime: node.birthtime,
+    changeTime: node.ctime,
+    extension: node.ext.slice(1).toLowerCase(),
+    internal: stats.isDirectory()
+      ? {
+          contentDigest: crypto
+            .createHash(`md5`)
+            .update(
+              JSON.stringify({
+                absolutePath: node.absolutePath,
+                stats: stats,
+              })
+            )
+            .digest(`hex`),
+          type: `Directory`,
+        }
+      : {
+          contentDigest: await md5File(node.absolutePath),
+          mediaType: mime.getType(node.ext),
+          type: `File`,
+        },
+    modifiedTime: node.mtime,
+    prettySize: prettyBytes(node.size),
+    relativeDirectory: path.relative(node.cwd, node.dir),
   }
 
-  const stats = await fs.stat(slashedFile.absolutePath)
-  let internal
-  if (stats.isDirectory()) {
-    const contentDigest = crypto
-      .createHash(`md5`)
-      .update(
-        JSON.stringify({ stats: stats, absolutePath: slashedFile.absolutePath })
-      )
-      .digest(`hex`)
-    internal = {
-      contentDigest,
-      type: `Directory`,
-    }
-  } else {
-    const contentDigest = await md5File(slashedFile.absolutePath)
-    internal = {
-      contentDigest,
-      mediaType: mime.lookup(slashedFile.ext),
-      type: `File`,
-    }
+  node = {
+    ...node,
+    allFiles: stats.isDirectory()
+      ? await fs
+          .readdir(node.absolutePath)
+          .map(file => path.join(node.absolutePath, file))
+          .filter(file => fs.statSync(file).isFile())
+          .map(file => createFileNode(file, opts))
```

Comment:
Instead of creating new file duplicate file nodes — you should instead link to them using the `___node` syntax.

Reply:
Is there documentation for the ___node syntax? I haven’t come across any mention of that before now.